src/RadixDlt.NetworkGateway.PostgresIntegration/Models/EntityResourceVaultAggregateHistory.cs

```diff
@@ -84,6 +89,107 @@ internal class EntityResourceVaultAggregateHistory
     [Column("resource_entity_id")]
     public long ResourceEntityId { get; set; }
 
+    /// <summary>
+    /// The aggregate list of all vaults ordered by most recently updated.
+    /// </summary>
     [Column("vault_entity_ids")]
     public List<long> VaultEntityIds { get; set; }
+
+    /// <summary>
+    /// The last significant update state version for resources stored in <see cref="VaultEntityIds"/>.
+    /// </summary>
+    /// <remarks>
+    /// This property is nullable in order to ensure smooth database migration process.
+    /// </remarks>
+    [Column("vault_significant_update_state_versions")]
+    public List<long>? VaultSignificantUpdateStateVersions { get; set; }
+
+    public static EntityResourceVaultAggregateHistory Create(long databaseId, long entityId, long resourceEntityId, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, null, entityId, resourceEntityId, stateVersion);
+    }
+
+    public static EntityResourceVaultAggregateHistory CopyOf(long databaseId, EntityResourceVaultAggregateHistory other, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, other, other.EntityId, other.ResourceEntityId, stateVersion);
+    }
+
+    /// <summary>
+    /// Attempts to add new or update existing vault unless it already is the most recently modified one.
+    /// </summary>
+    /// <returns>true if added or modified, otherwise false.</returns>
+    public bool TryUpsert(long vaultId, long stateVersion)
+    {
+        EnsureVaultSignificantUpdateStateVersionsExists();
+
+        var currentIndex = VaultEntityIds.IndexOf(vaultId);
+
+        // we're already the most recent one and there's no point potential second resource with the same state version
+        if (currentIndex == 0 && VaultEntityIds.Count == 1)
+        {
+            return false;
+        }
+
+        // we're already the most recent one but there might be second resource with matching state_version
+        // so in order to guarantee that we'll be unambiguously earlier in the overall sequence we might potentially
+        // want to update anyways
+        if (currentIndex == 0 && VaultEntityIds.Count > 1 && VaultSignificantUpdateStateVersions[1] < VaultSignificantUpdateStateVersions[0])
+        {
+            return false;
+        }
+
+        if (currentIndex != -1)
+        {
+            VaultEntityIds.RemoveAt(currentIndex);
+            VaultSignificantUpdateStateVersions.RemoveAt(currentIndex);
+        }
+
+        VaultEntityIds.Insert(0, vaultId);
+        VaultSignificantUpdateStateVersions.Insert(0, stateVersion);
+
+        return true;
+    }
+
+    public bool ShouldBePersisted()
+    {
+        if (_originalVaultEntityIds == null)
+        {
+            return true;
+        }
+
+        if (!_originalVaultEntityIds.SequenceEqual(VaultEntityIds.ToArray()))
```

Comment:
It feels like we should also check for a change in `VaultSignificantUpdateStateVersions`?